Add package tiny-vcdiff

jue89 opened the pull request:
### Contribution description

This is a WIP PR that brings [tiny-vcdiff](https://github.com/jue89/tiny-vcdiff) into RIOT.
I mentioned this library during [my talk on the last RIOT Summit](https://www.youtube.com/watch?v=fs69B8rLcMI&t=1464s)

Currently this PR supports `mtd` as backend. But a `vfs` will be added, too. I guess those are the obvious interfaces for such a package.

### Testing procedure

I will add a test later, that will make use of all implemented backends.

### Issues/PRs references

aabadie commented on pkg/tinyvcdiff/doc.txt:
```diff
@@ -0,0 +1,15 @@
+ *
+ * # Introduction
+ *
+ * Put an introduction of the package here
```
This generated line needs to be updated ;)

jue89 replied:
I've written some lines how to use this lib.

aabadie commented on pkg/tinyvcdiff/include/vcdiff_mtd.h:
```diff
@@ -0,0 +1,52 @@
+ * @{
+ *
+ * @file
+ * @defgroup pkg_tinyvcdiff   Tiny VCDIFF
+ * @ingroup  pkg
```
Duplicated from doc.txt (maybe a bug in the generator ? )

```suggestion
 * @ingroup  pkg_tinyvcdiff
```

jue89 replied:
Good catch! Thank you. Yes, too much copy pasta involved ;-)

aabadie commented:
I'm realizing that there's no test application for this new package. So it won't be built by the CI. I would be nice to add a simple one (you can use `make generate-test`)

jue89 commented:
> I'm realizing that there's no test application for this new package. So it won't be built by the CI. I would be nice to add a simple one (you can use `make generate-test`)

Yep, I'm currently working on that. The reason why this is a draft PR.

jue89 commented:
Writing the test revealed the bugs described in #17865 and #17866. Waiting for them to be on master.

benpicco commented:
You can probably squash when you rebase

jue89 commented:
I think everything should be in place, now. I ready to be ~~roasted~~ reviewed :see_no_evil:

benpicco commented on pkg/tinyvcdiff/contrib/tinyvcdiff_mtd/mtd.c:
```diff
@@ -0,0 +1,55 @@
+#include "vcdiff_mtd.h"
```
The file needs the usual boilerplate header

benpicco commented on pkg/tinyvcdiff/contrib/tinyvcdiff_mtd/mtd.c:
```diff
@@ -0,0 +1,55 @@
+    return rc;
+}
+
+static int read (void *dev, uint8_t *dest, size_t offset, size_t len)
```
```suggestion
static int _read(void *dev, uint8_t *dest, size_t offset, size_t len)
```

You want to avoid clashing with [`read(2)`](https://man7.org/linux/man-pages/man2/read.2.html)

benpicco commented on pkg/tinyvcdiff/contrib/tinyvcdiff_vfs/vfs.c:
```diff
@@ -0,0 +1,66 @@
+#include "vcdiff_vfs.h"
+#include "errno.h"
+
+static int read (void *dev, uint8_t *dest, size_t offset, size_t len)
```
```suggestion
static int _read(void *dev, uint8_t *dest, size_t offset, size_t len)
```

Now you *really* don't want to clash with [`read(2)`](https://man7.org/linux/man-pages/man2/read.2.html)

benpicco commented on pkg/tinyvcdiff/contrib/tinyvcdiff_vfs/vfs.c:
```diff
@@ -0,0 +1,66 @@
+static int read (void *dev, uint8_t *dest, size_t offset, size_t len)
+{
+    int rc;
+    vcdiff_vfs_t *vfs = (vcdiff_vfs_t *) dev;
```
```suggestion
    vcdiff_vfs_t *vfs = dev;
```

no need to cast the `void *`

benpicco commented on pkg/tinyvcdiff/contrib/tinyvcdiff_vfs/vfs.c:
```diff
@@ -0,0 +1,66 @@
+    /* write until all bytes are stored in the buffer */
+    offset = 0;
+    while (len) {
+        rc = vfs_read(vfs->fd, &dest[offset], len);
+        if (rc < 0) {
+            return rc;
+        }
+        len -= rc;
+        offset += rc;
```
Shouldn't it be?
```suggestion
        rc = vfs_read(vfs->fd, dest, len);
        if (rc < 0) {
            return rc;
        }
        len -= rc;
        dest += rc;
```

Or is the `offset` really *both* into the destination buffer *and* the source file?

jue89 replied:
No, just the offset in the file, not the buffer. I'm reusing the variable and set it to `0` before entering the loop.

But I guess moving the pointer is also a common way to go that is understood by most programmers ;-) Changed it.

50a48a8a489417608b095e44c198e9b331c574fd

benpicco commented on tests/pkg_tinyvcdiff/main.c:
```diff
@@ -0,0 +1,122 @@
+        .parent = &mtd_parent,
+        .sector = 0
+    };
+    TEST_ASSERT_EQUAL_INT(0, mtd_write(&mtd_source.mtd, (uint8_t*)bin_source, 0, strlen(bin_source)));
```
```suggestion
    TEST_ASSERT_EQUAL_INT(0, mtd_write(&mtd_source.mtd, bin_source, 0, sizeof(bin_source)));
```

jue89 replied:
`sizeof(bin_source) != strlen(bin_source)`. The binary delta doesn't account for the trailing `\0`.

benpicco commented on tests/pkg_tinyvcdiff/main.c:
```diff
@@ -0,0 +1,122 @@
+    };
+    TEST_ASSERT_EQUAL_INT(0, mtd_write(&mtd_source.mtd, (uint8_t*)bin_source, 0, strlen(bin_source)));
+    vcdiff_mtd_t vcdiff_source = VCDIFF_MTD_INIT(&mtd_source.mtd);
+    vcdiff_set_source_driver(&vcdiff, &vcdiff_mtd_driver, (void*)&vcdiff_source);
```
```suggestion
    vcdiff_set_source_driver(&vcdiff, &vcdiff_mtd_driver, &vcdiff_source);
```
should work too since `vcdiff_set_source_driver` takes a `void *`

benpicco commented on tests/pkg_tinyvcdiff/main.c:
```diff
@@ -0,0 +1,122 @@
+        .sector = 0
+    };
+    TEST_ASSERT_EQUAL_INT(0, mtd_write(&mtd_source.mtd, (uint8_t*)bin_source, 0, strlen(bin_source)));
+    vcdiff_mtd_t vcdiff_source = VCDIFF_MTD_INIT(&mtd_source.mtd);
```
not sure if this is really better than 
```suggestion
    vcdiff_mtd_t vcdiff_source = { .dev = &mtd_source.mtd };
```

jue89 replied:
I tried to pick up the static init that I learned from Mutex, ... ;-)

I like that it's easier to stay compatible if that struct changes in the future.

benpicco commented on tests/pkg_tinyvcdiff/main.c:
```diff
@@ -0,0 +1,122 @@
+    TEST_ASSERT_EQUAL_INT(0, vfs_format(&vfs_target));
+    TEST_ASSERT_EQUAL_INT(0, vfs_mount(&vfs_target));
+    int fd_target = vfs_open("/fatfs/target", O_WRONLY | O_CREAT, 0);
+    vcdiff_vfs_t vcdiff_target = VCDIFF_VFS_INIT(fd_target);
```
Please keep the VFS test in a separate function, mixing them like this is confusing.

jue89 replied:
I split it up into two tests. Furthermore, I remove `mtd_mapper` and built a fake MTD driver.

Has this improved the situation?

benpicco commented on tests/pkg_tinyvcdiff/main.c:
```diff
@@ -0,0 +1,122 @@
+static const uint8_t diff[] = {
+    0xd6, 0xc3, 0xc4, 0x53, 0x00, 0x01, 0x2b, 0x00, 0x17, 0x2e, 0x00, 0x00,
+    0x12, 0x00, 0x0f, 0x48, 0x65, 0x6c, 0x6c, 0x6f, 0x20, 0x75, 0x6e, 0x69,
+    0x76, 0x65, 0x72, 0x73, 0x65, 0x13, 0x20, 0x0b
```
How was this binary data created?

jue89 replied:
Added a comment explaining how I generated this binary data.

d226cee446b1dec2b8bdd424511df4ce3950c913

benpicco commented on tests/pkg_tinyvcdiff/fakemtd.h:
```diff
@@ -0,0 +1,77 @@
+/**
+ * @brief Sector count
+ *
+ * @note FatFS requires more than 128 sectors
```
Then let's use something else so we have a chance of running this on some targets too.

<details><summary>littlefs2</summary>

```patch
diff --git a/tests/pkg_tinyvcdiff/Makefile b/tests/pkg_tinyvcdiff/Makefile
index 9667e2b9c9..59bb9b1abd 100644
--- a/tests/pkg_tinyvcdiff/Makefile
+++ b/tests/pkg_tinyvcdiff/Makefile
@@ -12,7 +12,7 @@ USEPKG += tinyvcdiff
 USEMODULE += mtd
 
 # VFS backend
-USEPKG += fatfs
+USEPKG += littlefs2
 USEMODULE += fatfs_vfs
 USEMODULE += fatfs_vfs_format
 
diff --git a/tests/pkg_tinyvcdiff/fakemtd.h b/tests/pkg_tinyvcdiff/fakemtd.h
index 1c3a0d8302..e6a20f9e70 100644
--- a/tests/pkg_tinyvcdiff/fakemtd.h
+++ b/tests/pkg_tinyvcdiff/fakemtd.h
@@ -31,17 +31,13 @@ extern "C" {
 
 /**
  * @brief Pages per sector
- *
- * @note FAKE_MTD_PAGE_SIZE * FAKE_MTD_PAGES_PER_SECTOR must be 512 for FatFS
  */
 #define FAKE_MTD_PAGES_PER_SECTOR 4
 
 /**
  * @brief Sector count
- *
- * @note FatFS requires more than 128 sectors
  */
-#define FAKE_MTD_SECTOR_COUNT     256
+#define FAKE_MTD_SECTOR_COUNT     4
 
 /**
  * @brief Storage size in byte
diff --git a/tests/pkg_tinyvcdiff/main.c b/tests/pkg_tinyvcdiff/main.c
index d2161db690..7b5c859fbd 100644
--- a/tests/pkg_tinyvcdiff/main.c
+++ b/tests/pkg_tinyvcdiff/main.c
@@ -26,7 +26,7 @@
 #include "vcdiff_mtd.h"
 #include "vcdiff_vfs.h"
 #include "fakemtd.h"
-#include "fs/fatfs.h"
+#include "fs/littlefs2_fs.h"
 
 static const char *source_bin = "Hello world! I hope you are doing well ...\n";
 static const char *target_bin = "Hello universe! I hope you are doing well ...\n";
@@ -82,10 +82,10 @@ static void test_tinyvcdiff_vfs(void)
     static vcdiff_t vcdiff;
 
     static fake_mtd_t storage = FAKE_MTD_INIT;
-    fatfs_desc_t fs = { .dev = &storage.mtd };
+    littlefs2_desc_t fs = { .dev = &storage.mtd };
     vfs_mount_t mnt = {
         .mount_point = "/mnt",
-        .fs = &fatfs_file_system,
+        .fs = &littlefs2_file_system,
         .private_data = &fs,
     };
     vfs_format(&mnt);
```
</details>

I'm happy to report that this then also works on AVR.

jue89 replied:
Nice! Didn't expected that - the mandatory address cache needs some kB of space in RAM.

jue89 replied:
156b8dd9c125ef150a963216537b063d62753456

And I shared the large context objects between both test cases: 375b241933fd93b8715e58b5e7d1ddb7b8079903

benpicco reviewed (commented):
Nice, works like a charm - some more nits

jue89 commented on tests/pkg_tinyvcdiff/main.c:
```diff
@@ -0,0 +1,146 @@
+    /* setup vcdiff */
+    vcdiff_init(&vcdiff);
+    vcdiff_set_source_driver(&vcdiff, &vcdiff_mtd_driver, &source_vcdiff);
+    vcdiff_set_target_driver(&vcdiff, &vcdiff_mtd_driver, &target_vcdiff);
```
Our use case is:

* The currently running Riotboot slot is mapped as mtd and used as source
* The binary delta is transmitted in chunks over LoRaWAN
* The resulting slot is then written on an external flash device
* The system is rebooted after successful integrity check and Riotboot copies over the new slot into place.

fjmolinas commented on tests/pkg_tinyvcdiff/app.config.test:
```diff
@@ -0,0 +1,6 @@
+CONFIG_MODULE_EMBUNIT=y
+CONFIG_PACKAGE_TINYVCDIFF=y
+CONFIG_MODULE_MTD=y
+CONFIG_PACKAGE_FATFS=y
+CONFIG_MODULE_FATFS_VFS=y
+CONFIG_MODULE_FATFS_VFS_FORMAT=y
```
`FAT` does not seem to be yet moduled in Kconfig :(

jue89 replied:
It got replaced by littlefs2 - so that problem has disappeared for my PR :D

jue89 replied:
Well ... littlefs2 isn't, too.

I'll remove it from the `app.config.test`, right?

fjmolinas replied:
Yes, vfs and filesystems are not yet modelled...

jue89 replied:
b37095f26145fd45da9c83534b40e90716a4b719

jue89 replied:
Okay, I see ... we cannot have just one half of the story. I have to model `littlfs2` in Kconfig before I can use Kconfig for this PR, right?

fjmolinas commented on tests/pkg_tinyvcdiff/fakemtd.h:
```diff
@@ -0,0 +1,77 @@
+/**
+ * @brief Storage size in byte
+ */
+#define MAKE_MTD_MEM_SIZE         FAKE_MTD_PAGE_SIZE * FAKE_MTD_PAGES_PER_SECTOR * FAKE_MTD_SECTOR_COUNT
```
```suggestion
#define MAKE_MTD_MEM_SIZE         (FAKE_MTD_PAGE_SIZE * FAKE_MTD_PAGES_PER_SECTOR * FAKE_MTD_SECTOR_COUNT)
```

Also this overflows.

fjmolinas reviewed (commented):
Seems CI is mostly happy, but this will need some blacklisting

jue89 commented:
Thank you for your reviews, @benpicco @fjmolinas!

Is it okay if I squash all fixups before addressing your suggestions?

benpicco commented:
Sure, feel free to squash

jue89 commented on tests/pkg_tinyvcdiff/main.c:
```diff
@@ -0,0 +1,146 @@
+    vcdiff_mtd_t target_vcdiff = VCDIFF_MTD_INIT(target_mtd);
+
+    /* write source data to fake-mtd */
+    mtd_write(source_mtd, source_bin, 0, strlen(source_bin));
```
d298e17bfa0f9b13f20106f7d928e81f49293e31

jue89 commented on pkg/tinyvcdiff/contrib/tinyvcdiff_vfs/vfs.c:
```diff
@@ -0,0 +1,74 @@
+            return rc;
+        }
+        len -= (size_t) rc;
+        offset += (size_t) rc;
```
3abd7f0c5d66f281803943c702b229fedbefefb9

jue89 commented on pkg/tinyvcdiff/contrib/tinyvcdiff_mtd/mtd.c:
```diff
@@ -0,0 +1,63 @@
+
+    /* early exit if all sectors are already in the erased state */
+    if (cnt == 0) {
+        goto ret;
```
4d4fe099247644eb96c4d7dabaff9ed07faf1a40

jue89 commented on pkg/tinyvcdiff/contrib/tinyvcdiff_mtd/mtd.c:
```diff
@@ -0,0 +1,63 @@
+    /* erase sectors */
+    rc = mtd_erase_sector(mtd->dev, mtd->next_erase_sector, cnt);
+    if (rc != 0) {
+        goto ret;
```
4d4fe099247644eb96c4d7dabaff9ed07faf1a40

benpicco commented:
If you have all the toolchains installed you can create `Makefile.ci` by running 

    make create-Makefile.ci

benpicco commented on pkg/tinyvcdiff/contrib/tinyvcdiff_vfs/vfs.c:
```diff
@@ -0,0 +1,73 @@
+    }
+
+    /* write until all bytes are stored in the buffer */
+    while (len) {
+        rc = vfs_read(vfs->fd, dest, len);
+        if (rc < 0) {
+            return rc;
+        }
+        len -= rc;
+        dest += rc;
+    }
```
tbh I'm wondering why this is needed in the first place.
If the underlying filesystem driver would return less bytes than requested with the expectation that you call read again, that should be handled inside `vfs_read()`.

But normally if you read less bytes than you requested, that means the file just ended.

jue89 replied:
Good point. At least I should break if `rc == 0` and return an error.

Aside from that, I'm not sure if `vfs_read()` is guaranteed to read all available data in one go. That's the reason for the loop. Do you have a hint if this is too careful?

benpicco replied:
> I'm not sure if `vfs_read()` is guaranteed to read all available data in one go.

It should because it can grantee that.

jue89 commented:
Rebased on latest master.

jue89 commented on pkg/tinyvcdiff/doc.txt:
```diff
@@ -0,0 +1,58 @@
+ *     vcdiff_vfs_t target = VCDIFF_VFS_INIT(fd_target);
+ *     vcdiff_set_target_driver(&vcdiff, &vcdiff_vfs_driver, &target);
+ *
+ *     rc = vcdiff_apply_delta(&vcdiff, data, data_len);
```
Nope, It should work byte-wise. Like show [here](https://github.com/jue89/tiny-vcdiff/blob/d275c7d1f3fff71c5882f80434a12169a32a33f9/tests/vcdiff.c#L265-L267).

If it doesn't, it's a bug ;-)

jue89 commented:
Well, I could. But that produces really noisy code :-/

fjmolinas commented on pkg/tinyvcdiff/contrib/tinyvcdiff_mtd/mtd.c:
```diff
@@ -0,0 +1,62 @@
+    /* mtd_read_page will take care of calculation the right page
+     * it's safe to call mtd_read_page_raw: the erase driver already
+     * prepared the sector for writing */
+    int rc = mtd_write_page_raw(mtd->dev, src, 0, offset, len);
```
Are you sure this is right? The input `offset` is offset in bytes, while the offset parameter to `mtd_write_page_raw` is the offset in the page being written to, shouldn't it be something like:

```suggestion
    uint32_t page = offset / mtd->dev->page_size;
    uint32_t page_offset = (offset % mtd->dev->page_size);
    int rc = mtd_write_page_raw(mtd->dev, src, page, page_offset, len);
```

(untested)

fjmolinas replied:
Same question goes for the read function actually.

jue89 replied:
It should work to start at page `0` and let the MTD implemenation [figure out the right page](https://github.com/RIOT-OS/RIOT/blob/fca56ba0c3383f713179546e1a3a9b5bc2f24fd8/drivers/mtd/mtd.c#L255-L256).

jue89 replied:
Well, but another problem could occur: un-aligned writes (i.e. `addr % FLASHPAGE_WRITE_BLOCK_ALIGNMENT` and `size % FLASHPAGE_WRITE_BLOCK_SIZE` are possible troublemaker ...). Or am I wrong?

benpicco replied:
yea but you don't get those through MTD - #17683 tries to fix that

jue89 replied:
I added write alignment for the MTD backend: b705803

jue89 commented:
I think it would be nice to have working MTD ;-) Thus, I'm waiting for #17683.

jue89 commented:
I added write alignment for this package. Once #17683 is merged, I'll add a check if the configured write alignment and MTD write alignment are compatible.

jue89 commented:
> Can you make the fall-throughs explicit with `/* fall-through */`?

I sneaked that in with some pre-processor magic. The package can now be compiled without `-Wno-implicit-fallthrough`.

jue89 commented:
Finally a happy Murdock :)

jue89 commented:
Rebased on latest master. Still looking fine.

@fjmolinas You faced problems with alignment on MTD flashpage, right? Could you test this PR again? I've introduced aligned writes for MTD backends.

benpicco reviewed (commented):
With a rebase you can now get rid of `CONFIG_TINYVCDIFF_MTD_WRITE_SIZE`

MrKevinWeiss commented:
It seems like `native` tests are failing now hitting an [assert](https://github.com/RIOT-OS/RIOT/blob/master/drivers/mtd/mtd.c#L53)

jue89 commented:
Oh thank you for your review!

jue89 commented on pkg/tinyvcdiff/contrib/tinyvcdiff_mtd/mtd.c:
```diff
@@ -0,0 +1,179 @@
+    size_t alignment_offset;
+    size_t copy_len;
+
+    alignment_offset = mtd->offset % CONFIG_TINYVCDIFF_MTD_WRITE_SIZE;
```
`vcdiff_mtd_t` requires `write_size` to be known at compile time to adjust buffer sizes. But at least an `assert()` would be nice to have.

benpicco replied:
Could we use `mtd->work_area` instead?

jue89 replied:
This would add the dependency `mtd_write_page`. Currently, I'm not using any of the write page features. Adding this module would add unnecessary code just to have the (`mtd_write_page`-internal) buffer at head. It feels a little bit hacky to me.

I'd say having `assert()`s in place that make sure the buffer is aligned and large enough would be a good thing to have in place.

The default of 4 bytes seems to be a size that fits all `mtd` devices available in RIOT (IIRC) and isn't wasteful in my opinion.

benpicco replied:
For `mtd_flashpage` it can be larger than 4, same for SD cards (but those are usually used with a filesystem)
